core/arch/arm/plat-rockchip/psci_rk322x.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2017, Fuzhou Rockchip Electronics Co., Ltd.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <cru.h>
+#include <console.h>
+#include <delay.h>
+#include <initcall.h>
+#include <io.h>
+#include <kernel/generic_boot.h>
+#include <kernel/misc.h>
+#include <mm/core_mmu.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <sm/optee_smc.h>
+#include <sm/psci.h>
+#include <stdint.h>
+#include <tee/entry_std.h>
+#include <tee/entry_fast.h>
+
+#define CORE_WFE_MSK(core)	SHIFT_U32(0x02, (core))
+#define CORE_WFI_MSK(core)	SHIFT_U32(0x20, (core))
+#define CORE_WFE_I_MSK(core)	(CORE_WFI_MSK(core) | CORE_WFE_MSK(core))
+
+#define LOCK_TAG		0xDEADBEAF
+#define LOCK_ADDR_OFFSET	4
+#define BOOT_ADDR_OFFSET	8
+#define GRF_CPU_STATUS1		0x524
+
+static void wait_core_wfe_i(uint32_t core)
+{
+	uint32_t wfei_mask, loop = 0;
+	vaddr_t va_base = (vaddr_t)phys_to_virt_io(GRF_BASE);
+
+	wfei_mask = CORE_WFE_I_MSK(core);
+	while (!(read32(va_base + GRF_CPU_STATUS1) & wfei_mask) && loop < 500) {
```

Comment:
I discussed with my team mate, we thought it is better to print error message and return an error code, then give up cpu up this time.